### PR TITLE
feat(aws): Adds support for Cloudwatch dashboards & metric alarms.  

### PR DESCRIPTION
**Pull request by robh007**
**Objective**:

Adds support for cloudwatch_dashboard & cloudwatch_metric_alarm.

** Closes ** #54 

**Example output**: 

```
NAME                                              MONTHLY QTY  UNIT     PRICE   HOURLY COST  MONTHLY COST  

  aws_cloudwatch_dashboard.main                                                                              
  └─ Dashboard                                                1  monthly  3.0000       0.0041        3.0000  
  Total                                                                                0.0041        3.0000  
                                                                                                             
  aws_cloudwatch_metric_alarm.expression                                                                     
  └─ Standard resolution                                      1  alarm    0.1000       0.0001        0.1000  
  Total                                                                                0.0001        0.1000  
                                                                                                             
  aws_cloudwatch_metric_alarm.high                                                                           
  └─ High resolution                                          1  alarm    0.3000       0.0004        0.3000  
  Total                                                                                0.0004        0.3000  
                                                                                                             
  aws_cloudwatch_metric_alarm.standard                                                                       
  └─ Standard resolution                                      1  alarm    0.1000       0.0001        0.1000  
  Total                                                                                0.0001        0.1000  
                                                                                                             
  aws_cloudwatch_metric_alarm.xx_anomaly_detection                                                           
  └─ Standard resolution anomaly detection                    3  alarm    0.1000       0.0004        0.3000  
  Total                                                                                0.0004        0.3000  
                                                                                                             
  OVERALL TOTAL (USD)                                                                  0.0052        3.8000  
```
**Pricing details**:
The Cloudwatch service provides a single pricing component for dashboards & 5 pricing components for alarms. Composite alarms are not currently supported by terraform. Pricing details describe here https://aws.amazon.com/cloudwatch/pricing/

### CODE REVIEW

File: internal/providers/terraform/aws/cloudwatch_dashboard.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+	}
+}
+
+func NewCloudwatchDashboard(d *schema.ResourceData, u *schema.ResourceData) *schema.Resource {
+	return &schema.Resource{
+		Name: d.Address,
+		CostComponents: []*schema.CostComponent{
+			{
+				Name:            "Dashboard",
+				Unit:            "monthly",
```
Comment (aliscott) [inline internal/providers/terraform/aws/cloudwatch_dashboard.go:21]:
```suggestion
				Unit:            "months",
```

**Review by aliscott — COMMENTED** [review]
@robh007 thank you! 😄 

I've been going through the CloudWatch prices and I think we should also be covering the Log pricing, without usage intially. From what I can tell the following prices can be added to these resources:

`aws_cloudwatch_log_group`:
* Collect (Data Ingestion)
* Store (Archival)
* Analyze (Logs Insights queries)

`aws_cloudwatch_log_destination`:
* Delivery to CloudWatch Logs
* Delivery to S3
* Delivery to Kinesis Data Firehose

For these ones we can maybe detect the type of the `target_arn` and show the matching price.

---

Just for information, here's what I found for the 7 tabs of CloudWatch pricing (https://aws.amazon.com/cloudwatch/pricing/):

1. Metrics - we cover this in the EC2 resources currently - we might need to look at moving this to its own resource, but will look at that more when we do Data Transfer costs as it is a similar shared cost.
2. Dashboards - covered with this PR
3. Alarms - covered with this PR
4. Logs - we can cover these (as above)
5. Events - these are covered in the "EventBridge (CloudWatch Events)" Terraform resources, so 2w
6. Contributor insights - not supported yet by Terraform
7. Canaries - not supported yet by Terraform

File: internal/providers/terraform/aws/cloudwatch_metric_alarm.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+	var anomalyDetection string
+
+	quantity = decimal.NewFromInt(1)
+
+	if checkAnomlyDetection(d) {
+		quantity = decimal.NewFromInt(3)
+		anomalyDetection = " anomaly detection"
+	}
+
+	for _, query := range d.Get("metric_query").Array() {
```
[the remark this answers is not in the document]
Comment (robh007) [reply internal/providers/terraform/aws/cloudwatch_metric_alarm.go:64]: @aliscott the period within the metric if I'm reading the docs correctly needs to be the same across all metrics. When using metric_query you can query up to 10 metrics. All I can find is that metric periods need to all be the same within a metric query array. So it's the total within the metric query that needs adding to the calculation I think.

TF returns a validation error shown below. 

```
Creating metric alarm failed: ValidationError: All metrics in the alarm should have the same period
```

[the remark this answers is not in the document]
Comment (aliscott) [reply internal/providers/terraform/aws/cloudwatch_metric_alarm.go:64]: 👍 @robh007 interesting, thanks for investigating.

**Comment by robh007** [comment]
Added ```cloudwatch_log_group```

```
 aws_cloudwatch_log_group.logs                                                                                    
  ├─ Collect (Data Ingestion)                                 0  GB             0.5000       0.0000        0.0000  
  └─ Store (Archival)                                         0  GB             0.0300       0.0000        0.0000  
  Total                                                                                      0.0000        0.0000  
```

File: internal/providers/terraform/aws/cloudwatch_log_group.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+		gbDataAnalzyed = decimal.NewFromInt(u.Get("data_analyzed.0.value").Int())
+	}
+
+	if u != nil && u.Get("data_queries.0.value").Exists() {
+		gbDataAnalzyedQueries = u.Get("data_queries.0.value").Int()
+	}
+
+	costComponents := []*schema.CostComponent{
+		{
+			Name:            "Collect (Data Ingestion)",
```
Comment (aliscott) [inline internal/providers/terraform/aws/cloudwatch_log_group.go:42]:
```suggestion
			Name:            "Data ingestion",
```

File: internal/providers/terraform/aws/cloudwatch_log_group.go [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+				Region:        strPtr(region),
+				Service:       strPtr("AmazonCloudWatch"),
+				ProductFamily: strPtr("Data Payload"),
+				AttributeFilters: []*schema.AttributeFilter{
+					{Key: "usagetype", ValueRegex: strPtr("/-DataProcessing-Bytes/")},
+				},
+			},
+		},
+		{
+			Name:            "Store (Archival)",
```
Comment (aliscott) [inline internal/providers/terraform/aws/cloudwatch_log_group.go:57]:
```suggestion
			Name:            "Archival storage",
```

File: internal/providers/terraform/aws/cloudwatch_log_group.go [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+				AttributeFilters: []*schema.AttributeFilter{
+					{Key: "usagetype", ValueRegex: strPtr("/-TimedStorage-ByteHrs/")},
+				},
+			},
+		},
+	}
+
+	if gbDataAnalzyedQueries > 0 && gbDataAnalzyed.GreaterThan(decimal.NewFromInt(0)) {
+		costComponents = append(costComponents, &schema.CostComponent{
+			Name:            "Analyze (Logs insights queries)",
```
Comment (aliscott) [inline internal/providers/terraform/aws/cloudwatch_log_group.go:75]:
```suggestion
			Name:            "Insights queries data scanned",
```

File: internal/providers/terraform/aws/cloudwatch_log_group.go [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+			Unit:            "GB-scanned",
```
Comment (aliscott) [inline internal/providers/terraform/aws/cloudwatch_log_group.go:76]:
```suggestion
			Unit:            "GB",
```

File: internal/providers/terraform/aws/cloudwatch_log_group.go [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+				Service:       strPtr("AmazonCloudWatch"),
+				ProductFamily: strPtr("Storage Snapshot"),
+				AttributeFilters: []*schema.AttributeFilter{
+					{Key: "usagetype", ValueRegex: strPtr("/-TimedStorage-ByteHrs/")},
+				},
+			},
+		},
+	}
+
+	if gbDataAnalzyedQueries > 0 && gbDataAnalzyed.GreaterThan(decimal.NewFromInt(0)) {
```
Comment (aliscott) [inline internal/providers/terraform/aws/cloudwatch_log_group.go:73]: I talked with @alikhajeh1 about this as well and we agreed that we should always show the prices, even if there is no specified usage. In the future we will probably have a better way of filtering prices that are not relevant.

**Review by aliscott — COMMENTED** [review]
Looks good!

**Comment by robh007** [comment]
Thanks @aliscott I've updated & the output reads. 

```
 aws_cloudwatch_log_group.logs                                                                                    
  ├─ Data ingestion                                           -  GB             0.5000            -             -  
  ├─ Archival Storage                                         -  GB             0.0300            -             -  
  └─ Insights queries data scanned                            -  GB             0.0050            -             -  
  Total                                                                                           -             -  

```

**Pull request merged**